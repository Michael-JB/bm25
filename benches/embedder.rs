use bm25::{EmbedderBuilder, Language, LanguageMode, TokenEmbedder};
use divan::Bencher;
use std::hash::Hash;

use divan::AllocProfiler;

#[global_allocator]
static ALLOC: AllocProfiler = AllocProfiler::system();

fn main() {
    divan::main();
}

const ENGLISH_TEXT: &str =
    "The dominant sequence transduction models are based on complex recurrent or
convolutional neural networks in an encoder-decoder configuration. The best performing
models also connect the encoder and decoder through an attention mechanism. We
propose a new simple network architecture, the Transformer, based solely on attention
mechanisms, dispensing with recurrence and convolutions entirely. Experiments on two
machine translation tasks show these models to be superior in quality while being more
parallelizable and requiring significantly less time to train.";

#[divan::bench(args = [
    #[cfg(feature = "language_detection")]
    LanguageMode::Detect,
    LanguageMode::Fixed(Language::English),
], sample_count = 10000)]
fn language_mode(bencher: Bencher, language_mode: &LanguageMode) {
    let embedder =
        EmbedderBuilder::<u32>::with_fit_to_corpus(language_mode.clone(), &[ENGLISH_TEXT]).build();
    // Run once beforehand to warm up the cache
    embedder.embed(ENGLISH_TEXT);

    bencher.bench(|| embedder.embed(ENGLISH_TEXT));
}

#[divan::bench(types = [usize, u32, u64], sample_count = 10000)]
fn dimension<T>(bencher: Bencher)
where
    T: TokenEmbedder + Hash + Eq,
{
    bencher
        .with_inputs(|| {
            let embedder =
                EmbedderBuilder::<T>::with_fit_to_corpus(Language::English, &[ENGLISH_TEXT])
                    .build();
            // Run once beforehand to warm up the cache
            embedder.embed(ENGLISH_TEXT);
            embedder
        })
        .bench_values(|embedder| embedder.embed(ENGLISH_TEXT));
}

#[divan::bench]
fn fit_to_corpus(bencher: Bencher) {
    let corpus = [ENGLISH_TEXT; 5000];
    // Run once beforehand to warm up the cache
    let _ = EmbedderBuilder::<u32>::with_fit_to_corpus(Language::English, &corpus).build();

    bencher
        .bench(|| EmbedderBuilder::<u32>::with_fit_to_corpus(Language::English, &corpus).build());
}
