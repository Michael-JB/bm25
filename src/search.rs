use crate::embedder::{
    DefaultEmbeddingDimension, Embedder, EmbedderBuilder, EmbeddingDimension, LanguageMode,
};
use std::{
    collections::{HashMap, HashSet},
    fmt::{self, Debug, Display},
    hash::Hash,
    marker::PhantomData,
};

use crate::embedder::Embedding;

/// A search engine that ranks documents with BM25.
/// K is the type of the document id and D is the type of the embedding dimension.
pub struct SearchEngine<K, D = DefaultEmbeddingDimension>
where
    K: Eq + Hash + Clone,
    D: EmbeddingDimension,
{
    /// The embedder used to convert documents into embeddings.
    embedder: Embedder<D>,
    /// A mapping from document ids to the document embeddings.
    document_embeddings: HashMap<K, Embedding<D>>,
    /// A mapping from document ids to document contents.
    documents: HashMap<K, String>,
    /// A mapping from token indices to the number of documents that contain that token.
    term_frequencies: HashMap<D, u32>,
    /// A mapping from token indices to the set of documents that contain that token.
    inverted_term_index: HashMap<D, HashSet<K>>,
}

impl<K, D> Display for SearchEngine<K, D>
where
    K: Eq + Hash + Clone + Debug,
    D: EmbeddingDimension,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "SearchEngine {{ documents: {:?} }}", self.documents)
    }
}

/// A document that you can insert into a search engine. K is the type of the document, allowing
/// you to use any type as a document id. Note that it is more effient to use a numeric type.
#[derive(Eq, PartialEq, Debug, Clone, PartialOrd, Hash)]
pub struct Document<K>
where
    K: Eq + Hash + Clone + Debug,
{
    /// A unique identifier for the document.
    pub id: K,
    /// The contents of the document.
    pub contents: String,
}

impl<K> Display for Document<K>
where
    K: Eq + Hash + Clone + Debug,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.contents)
    }
}

impl<K> Document<K>
where
    K: Eq + Hash + Clone + Debug,
{
    /// Creates a new document with the given id and contents.
    pub fn new(id: K, contents: impl Into<String>) -> Document<K> {
        Document {
            id,
            contents: contents.into(),
        }
    }
}

/// A search result, containing a document and its BM25 score.
#[derive(PartialEq, Debug)]
pub struct SearchResult<K>
where
    K: Eq + Hash + Clone + Debug,
{
    /// The document that was found.
    pub document: Document<K>,
    /// The BM25 score of the document. A higher score means the document is more relevant to the
    /// query.
    pub score: f32,
}

impl<K, D> SearchEngine<K, D>
where
    K: Eq + Hash + Clone + Debug,
    D: EmbeddingDimension,
{
    /// Upserts a document into the search engine. If a document with the same id already exists,
    /// it will be replaced. Note that upserting a document will change the true value of `avgdl`.
    /// The more `avgdl` drifts from its true value, the less accurate the BM25 scores will be.
    pub fn upsert(&mut self, document: impl Into<Document<K>>) {
        let document = document.into();
        if self.documents.contains_key(&document.id) {
            self.remove(&document.id);
        }
        let Embedding { indices, values } = self.embedder.embed(document.contents.as_str());
        for token_index in indices.iter() {
            let term_frequency = self
                .term_frequencies
                .entry(token_index.clone())
                .or_insert(0);
            *term_frequency += 1;
            let document_set = self
                .inverted_term_index
                .entry(token_index.clone())
                .or_default();
            document_set.insert(document.id.clone());
        }
        self.document_embeddings
            .insert(document.id.clone(), Embedding { indices, values });
        self.documents
            .insert(document.id.clone(), document.contents);
    }

    /// Removes a document from the search engine if it exists.
    pub fn remove(&mut self, document_id: &K) {
        if let Some(document) = self.documents.remove(document_id) {
            let Embedding { indices, .. } = self.embedder.embed(&document);
            for token_index in indices.iter() {
                if let Some(term_frequency) = self.term_frequencies.get_mut(token_index) {
                    *term_frequency -= 1;
                }
                if let Some(document_set) = self.inverted_term_index.get_mut(token_index) {
                    document_set.remove(document_id);
                }
            }
            self.document_embeddings.remove(document_id);
        }
    }

    /// Gets the contents of a document by its id.
    pub fn get(&self, document_id: &K) -> Option<Document<K>> {
        self.documents.get(document_id).map(|contents| Document {
            id: document_id.clone(),
            contents: contents.clone(),
        })
    }

    fn idf(&self, token_index: &D) -> f32 {
        let term_frequency = *self.term_frequencies.get(token_index).unwrap_or(&0) as f32;
        let numerator = self.documents.len() as f32 - term_frequency + 0.5;
        let denominator = term_frequency + 0.5;
        (1f32 + (numerator / denominator)).ln()
    }

    /// Searches the documents for the given query and returns the top `limit` results.
    /// Only the document contents are searched, not the document ids.
    pub fn search(&self, query: &str, limit: impl Into<Option<usize>>) -> Vec<SearchResult<K>> {
        let Embedding { indices, .. } = self.embedder.embed(query);

        // Reduce search space by filtering out all documents whose score would be 0
        let relevent_embedding_it = indices
            .iter()
            .filter_map(|token_index| self.inverted_term_index.get(token_index))
            .flat_map(|document_set| document_set.iter())
            .collect::<HashSet<_>>()
            .into_iter()
            .filter_map(|document_id| {
                self.document_embeddings
                    .get(document_id)
                    .map(|embedding| (document_id, embedding))
            });

        // Compute the score of each candidate document
        let mut scores = HashMap::new();
        for (document_id, embedding) in relevent_embedding_it {
            let mut document_score = 0f32;
            for token_index in indices.iter() {
                let token_idf = self.idf(token_index);
                let token_index_value = embedding
                    .indices
                    .iter()
                    .zip(embedding.values.iter())
                    .find(|(i, _)| **i == *token_index)
                    .map(|(_, v)| *v)
                    .unwrap_or(0f32);
                let token_score = token_idf * token_index_value;
                document_score += token_score;
            }
            scores.insert(document_id, document_score);
        }

        // Sort and format results
        let mut results: Vec<_> = scores.iter().collect();
        results.sort_by(|a, b| b.1.partial_cmp(a.1).unwrap());
        results
            .into_iter()
            .take(limit.into().unwrap_or(usize::MAX))
            .filter_map(|(id, score)| {
                self.get(id).map(|document| SearchResult {
                    document,
                    score: *score,
                })
            })
            .collect()
    }
}

/// A consuming builder for SearchEngine.
pub struct SearchEngineBuilder<K, D = DefaultEmbeddingDimension>
where
    K: Eq + Hash + Clone + Debug,
    D: EmbeddingDimension,
{
    embedder_builder: EmbedderBuilder<D>,
    documents: Vec<Document<K>>,
    document_id_type: PhantomData<K>,
    embedding_dimension_type: PhantomData<D>,
}

impl<K, D> SearchEngineBuilder<K, D>
where
    K: Eq + Hash + Clone + Debug,
    D: EmbeddingDimension,
{
    /// Constructs a new SearchEngineBuilder with the given average document length. Use this if you
    /// know the average document length in advance. If you don't, but you have your full corpus
    /// ahead of time, use `with_documents` or `with_corpus` instead.
    ///
    /// If you have neither the full corpus nor a sample of it, you can configure the embedder to
    /// disregard document length by setting `b` to 0.0. In this case, it doesn't matter what
    /// value you pass to `with_avgdl`.
    ///
    /// The average document length is the average number of tokens in a document from your corpus;
    /// if you need access to this value, you can construct an Embedder and call `avgdl` on it.
    pub fn with_avgdl(avgdl: f32) -> SearchEngineBuilder<K, D> {
        SearchEngineBuilder {
            embedder_builder: EmbedderBuilder::<D>::with_avgdl(avgdl),
            documents: Vec::new(),
            document_id_type: PhantomData,
            embedding_dimension_type: PhantomData,
        }
    }

    /// Constructs a new SearchEngineBuilder with the given documents. The search engine will fit
    /// to the given documents, using the given language mode. When you call `build`, the builder
    /// will pre-populate the search engine with the given documents, and set its language mode to
    /// `language_mode`.
    pub fn with_documents(
        language_mode: impl Into<LanguageMode>,
        documents: impl IntoIterator<Item = impl Into<Document<K>>>,
    ) -> SearchEngineBuilder<K, D> {
        let documents = documents.into_iter().map(|d| d.into()).collect::<Vec<_>>();
        SearchEngineBuilder {
            embedder_builder: EmbedderBuilder::<D>::with_fit_to_corpus(
                language_mode,
                &documents
                    .iter()
                    .map(|d| d.contents.as_str())
                    .collect::<Vec<_>>(),
            ),
            documents,
            document_id_type: PhantomData,
            embedding_dimension_type: PhantomData,
        }
    }

    /// Constructs a new SearchEngineBuilder with the corpus. The search engine will fit
    /// to the given corpus, using the given language mode. When you call `build`, the builder
    /// will pre-populate the search engine with the given corpus. This function will
    /// automatically generate u32 ids for each entry in your corpus.
    pub fn with_corpus(
        language_mode: impl Into<LanguageMode>,
        corpus: impl IntoIterator<Item = impl Into<String>>,
    ) -> SearchEngineBuilder<u32, D> {
        let documents = corpus
            .into_iter()
            .enumerate()
            .map(|(id, document)| Document::new(id as u32, document.into()))
            .collect::<Vec<_>>();
        SearchEngineBuilder::<u32, D>::with_documents(language_mode.into(), documents)
    }

    /// Sets the language mode of the embedder.
    pub fn language_mode(mut self, language_mode: LanguageMode) -> SearchEngineBuilder<K, D> {
        self.embedder_builder.language_mode(language_mode);
        self
    }

    /// Sets the k1 parameter of the embedder.
    pub fn k1(mut self, k1: f32) -> SearchEngineBuilder<K, D> {
        self.embedder_builder.k1(k1);
        self
    }

    /// Sets the b parameter of the embedder.
    pub fn b(mut self, b: f32) -> SearchEngineBuilder<K, D> {
        self.embedder_builder.b(b);
        self
    }

    /// Overrides the average document length of the embedder.
    pub fn avgdl(mut self, avgdl: f32) -> SearchEngineBuilder<K, D> {
        self.embedder_builder.avgdl(avgdl);
        self
    }

    /// Builds the search engine.
    pub fn build(self) -> SearchEngine<K, D> {
        let mut search_engine = SearchEngine::<K, D> {
            embedder: self.embedder_builder.build(),
            document_embeddings: HashMap::new(),
            inverted_term_index: HashMap::new(),
            term_frequencies: HashMap::new(),
            documents: HashMap::new(),
        };
        for document in self.documents {
            search_engine.upsert(document);
        }
        search_engine
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        test_data_loader::tests::{read_recipes, Recipe},
        Language,
    };

    impl From<Recipe> for Document<String> {
        fn from(value: Recipe) -> Self {
            Document::new(value.title, value.recipe)
        }
    }

    fn create_recipe_search_engine(
        language_mode: impl Into<LanguageMode>,
    ) -> SearchEngine<String, u32> {
        let recipes = read_recipes("recipes_en.csv");

        SearchEngineBuilder::with_documents(language_mode, recipes).build()
    }

    #[test]
    fn search_returns_relevant_documents() {
        let corpus = vec!["space station", "bacon and avocado sandwich"];
        let search_engine =
            SearchEngineBuilder::<u32>::with_corpus(Language::English, corpus).build();

        let results = search_engine.search("sandwich with bacon", 5);
        assert!(results.len() == 1);
        assert!(results[0].document.contents == "bacon and avocado sandwich");
        assert!(results[0].score > 0.0);
    }

    #[test]
    fn search_does_not_return_unrelated_documents() {
        let corpus = vec!["space station", "bacon and avocado sandwich"];
        let search_engine =
            SearchEngineBuilder::<u32>::with_corpus(Language::English, corpus).build();

        let results = search_engine.search("maths and computer science", 5);
        assert!(results.is_empty());
    }

    #[test]
    fn it_can_insert_a_document() {
        let mut search_engine = SearchEngineBuilder::<&str>::with_avgdl(2.0).build();
        let document = Document::new("hello world", "bananas and apples");
        let document_id = document.id;

        search_engine.upsert(document.clone());
        let result = search_engine.get(&document_id);

        assert!(result.unwrap() == document);
    }

    #[test]
    fn it_can_remove_a_document() {
        let mut search_engine = SearchEngineBuilder::<usize>::with_avgdl(2.0).build();
        let document = Document::new(123, "bananas and apples");
        let document_id = document.id.clone();

        search_engine.upsert(document);
        search_engine.remove(&document_id);

        assert!(search_engine.get(&document_id).is_none());
    }

    #[test]
    fn it_can_update_a_document() {
        let document_id = "hello_world";
        let document = Document::new(document_id, "bananas and apples");
        let mut search_engine =
            SearchEngineBuilder::<&str>::with_documents(Language::English, vec![document]).build();
        let new_document = Document::new(document_id, "oranges and papayas");

        search_engine.upsert(new_document.clone());
        let result = search_engine.get(&document_id);

        assert!(result.unwrap() == new_document);
    }

    #[test]
    fn handles_empty_input() {
        let mut search_engine = SearchEngineBuilder::<u32>::with_avgdl(2.0).build();
        let document = Document::new(123, "");

        search_engine.upsert(document);

        let results = search_engine.search("bacon sandwich", 5);
        assert!(results.is_empty());
    }

    #[test]
    fn handles_empty_search() {
        let mut search_engine = SearchEngineBuilder::<u32>::with_avgdl(2.0).build();
        let document = Document::new(123, "pencil and paper");

        search_engine.upsert(document);

        let results = search_engine.search("", 5);
        assert!(results.is_empty());
    }

    #[test]
    fn it_returns_exact_matches_with_highest_score() {
        let search_engine = create_recipe_search_engine(LanguageMode::Fixed(Language::English));

        let results = search_engine.search(
            "To make guacamole, start by mashing 2 ripe avocados in a bowl.",
            None,
        );

        assert!(!results.is_empty());
        assert_eq!(results[0].document.id, "Guacamole");
    }

    #[test]
    fn it_only_returns_results_containing_query() {
        let search_engine = create_recipe_search_engine(Language::English);

        let results = search_engine.search("vegetable", 5);

        // At least 5 recipes contain the word "vegetable"
        assert_eq!(results.len(), 5);
        assert!(results
            .iter()
            .all(|result| result.document.contents.contains("vegetable")));
    }

    #[test]
    fn it_returns_results_sorted_by_score() {
        let search_engine = create_recipe_search_engine(Language::English);

        let results = search_engine.search("chicken", 1000);

        assert!(!results.is_empty());
        assert!(results
            .windows(2)
            .all(|result_pair| { result_pair[0].score >= result_pair[1].score }));
    }

    #[test]
    fn it_ranks_shorter_documents_higher() {
        let documents = [
            Document {
                id: 0,
                contents: "Correct horse battery staple bacon bacon bacon".to_string(),
            },
            Document {
                id: 1,
                contents: "Correct horse battery staple".to_string(),
            },
        ];
        let search_engine =
            SearchEngineBuilder::<u32>::with_documents(Language::English, documents).build();

        let results = search_engine.search("staple", 2);

        assert_eq!(results.len(), 2);
        assert_eq!(results[0].document.id, 1);
        assert_eq!(results[1].document.id, 0);
        assert!(results[0].score > results[1].score);
    }
}
