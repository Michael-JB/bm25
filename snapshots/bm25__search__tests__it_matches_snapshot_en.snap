---
source: src/search.rs
expression: results
---
[
    SearchResult {
        document: Document {
            id: "Apple Pie",
            contents: "To make apple pie, start by preheating your oven to 425°F (220°C). In a large bowl, combine 6 peeled and sliced apples with 1/2 cup of sugar, 1 teaspoon of cinnamon, and 1/4 teaspoon of nutmeg. Toss the apples to coat them evenly. Roll out a pie crust and fit it into a 9-inch pie dish. Pour the apple mixture into the crust and dot with 2 tablespoons of butter. Roll out the second pie crust and place it over the apples, crimping the edges to seal. Cut a few slits in the top crust to allow steam to escape. Bake for 45-50 minutes, or until the crust is golden brown and the filling is bubbly. Let the pie cool before serving.",
        },
        score: 1.0432872,
    },
    SearchResult {
        document: Document {
            id: "BBQ Chicken Wings",
            contents: "To make BBQ chicken wings, start by preheating your oven to 400°F (200°C). Place 2 pounds of chicken wings in a large bowl and toss them with 2 tablespoons of vegetable oil, salt, and pepper. Spread the wings out on a baking sheet lined with parchment paper. Bake in the preheated oven for 25-30 minutes, flipping halfway through, until the wings are golden brown and crispy. Remove the wings from the oven and toss them in your favorite BBQ sauce. Return the wings to the oven for an additional 5 minutes to allow the sauce to set. Serve hot, with extra sauce on the side.",
        },
        score: 1.515893,
    },
    SearchResult {
        document: Document {
            id: "Baked Ziti",
            contents: "To make baked ziti, preheat your oven to 375°F (190°C). Cook 1 pound of ziti pasta according to package instructions, then drain and set aside. In a large skillet, brown 1 pound of Italian sausage over medium heat, breaking it up into small pieces. Stir in 1 jar of marinara sauce and simmer for 5 minutes. In a large bowl, combine the cooked ziti with the sausage mixture, 1 cup of ricotta cheese, and 1/2 cup of grated Parmesan cheese. Pour the mixture into a greased baking dish and top with 2 cups of shredded mozzarella cheese. Bake for 25-30 minutes, or until the cheese is melted and bubbly. Let the ziti cool for a few minutes before serving.",
        },
        score: 1.6428902,
    },
    SearchResult {
        document: Document {
            id: "Banana Bread",
            contents: "To make banana bread, preheat your oven to 350°F (175°C). In a large bowl, mash 3 ripe bananas with a fork. Stir in 1/3 cup of melted butter, 1/2 cup of sugar, 1 beaten egg, and 1 teaspoon of vanilla extract. In a separate bowl, whisk together 1 1/2 cups of flour, 1 teaspoon of baking soda, and a pinch of salt. Gradually add the dry ingredients to the wet ingredients, stirring until just combined. Pour the batter into a greased loaf pan and bake for 60 minutes, or until a toothpick inserted into the center comes out clean. Let the bread cool in the pan for 10 minutes before transferring to a wire rack to cool completely.",
        },
        score: 1.4281813,
    },
    SearchResult {
        document: Document {
            id: "Blueberry Muffins",
            contents: "To make blueberry muffins, preheat your oven to 375°F (190°C) and line a muffin tin with paper liners. In a large bowl, whisk together 1 1/2 cups of flour, 3/4 cup of sugar, 1/2 teaspoon of salt, and 2 teaspoons of baking powder. In a separate bowl, combine 1/3 cup of vegetable oil, 1 large egg, and 1/3 cup of milk. Pour the wet ingredients into the dry ingredients and stir until just combined. Gently fold in 1 cup of fresh or frozen blueberries. Divide the batter evenly among the muffin cups and bake for 20-25 minutes, or until a toothpick inserted into the center comes out clean. Let the muffins cool in the tin for a few minutes before transferring to a wire rack to cool completely.",
        },
        score: 1.380263,
    },
    SearchResult {
        document: Document {
            id: "Chicken Parmesan",
            contents: "To make chicken Parmesan, preheat your oven to 400°F (200°C). Start by seasoning 2 boneless, skinless chicken breasts with salt and pepper. Dredge each chicken breast in flour, then dip in beaten eggs, and finally coat with breadcrumbs mixed with grated Parmesan cheese. In a large skillet, heat 2 tablespoons of olive oil over medium heat. Cook the chicken for 3-4 minutes on each side until golden brown. Transfer the chicken to a baking dish and top with marinara sauce and shredded mozzarella cheese. Bake for 15-20 minutes until the cheese is melted and bubbly. Serve hot over spaghetti or your favorite pasta.",
        },
        score: 1.4795465,
    },
    SearchResult {
        document: Document {
            id: "Chocolate Cake",
            contents: "To make a moist chocolate cake, preheat your oven to 350°F (175°C). Grease and flour two 9-inch round cake pans. In a large bowl, stir together 2 cups of sugar, 1 3/4 cups of flour, 3/4 cup of cocoa powder, 1 1/2 teaspoons of baking powder, 1 1/2 teaspoons of baking soda, and 1 teaspoon of salt. Add 2 eggs, 1 cup of milk, 1/2 cup of vegetable oil, and 2 teaspoons of vanilla extract. Beat the mixture on medium speed for 2 minutes. Stir in 1 cup of boiling water (the batter will be thin). Pour the batter evenly into the prepared pans. Bake for 30-35 minutes, or until a toothpick inserted into the center comes out clean. Cool the cakes in the pans for 10 minutes, then remove to wire racks to cool completely before frosting.",
        },
        score: 1.5382302,
    },
    SearchResult {
        document: Document {
            id: "Chocolate Chip Cookies",
            contents: "To make classic chocolate chip cookies, preheat your oven to 350°F (175°C). In a large bowl, cream together 1 cup of softened butter, 1 cup of white sugar, and 1 cup of brown sugar until smooth. Beat in 2 eggs, one at a time, then stir in 2 teaspoons of vanilla extract. In a separate bowl, whisk together 3 cups of flour, 1 teaspoon of baking soda, and 1/2 teaspoon of salt. Gradually blend the dry ingredients into the wet mixture. Fold in 2 cups of semisweet chocolate chips. Drop tablespoon-sized balls of dough onto ungreased baking sheets. Bake for 10-12 minutes, or until the edges are golden brown. Let the cookies cool on the baking sheet for a few minutes before transferring to a wire rack to cool completely.",
        },
        score: 1.7267127,
    },
    SearchResult {
        document: Document {
            id: "Eggplant Parmesan",
            contents: "To make eggplant Parmesan, start by preheating your oven to 375°F (190°C). Slice 2 medium eggplants into 1/4-inch thick rounds. Season the slices with salt and let them sit for 30 minutes to draw out excess moisture. Pat the eggplant slices dry with paper towels. Dredge each slice in flour, dip in beaten eggs, and coat with breadcrumbs. In a large skillet, heat 2 tablespoons of olive oil over medium heat and fry the eggplant slices until golden brown on both sides. In a baking dish, layer the fried eggplant slices with marinara sauce and shredded mozzarella cheese. Repeat the layers until all the ingredients are used, ending with a layer of cheese. Bake for 25-30 minutes until the cheese is melted and bubbly. Serve hot.",
        },
        score: 1.401157,
    },
    SearchResult {
        document: Document {
            id: "Garlic Bread",
            contents: "To make garlic bread, preheat your oven to 350°F (175°C). In a small bowl, mix together 1/2 cup of softened butter, 3 minced garlic cloves, and 2 tablespoons of chopped fresh parsley. Cut a loaf of French bread in half lengthwise. Spread the garlic butter mixture evenly over the cut sides of the bread. Place the bread halves on a baking sheet and bake for 10-12 minutes, or until the edges are golden and crispy. For extra cheesy garlic bread, sprinkle shredded mozzarella or Parmesan cheese over the top and bake for an additional 3-5 minutes, until the cheese is melted and bubbly. Slice and serve warm.",
        },
        score: 1.6725773,
    },
    SearchResult {
        document: Document {
            id: "Garlic Roasted Potatoes",
            contents: "To make garlic roasted potatoes, preheat your oven to 400°F (200°C). Cut four medium-sized potatoes into bite-sized cubes and place them in a large bowl. Toss the potatoes with three tablespoons of olive oil, four minced garlic cloves, salt, and pepper to taste. Spread the potatoes out in a single layer on a baking sheet. Roast in the preheated oven for 25-30 minutes, stirring halfway through, until the potatoes are golden brown and crispy. Garnish with chopped fresh parsley before serving.",
        },
        score: 1.1758184,
    },
    SearchResult {
        document: Document {
            id: "Lasagna",
            contents: "To make lasagna, preheat your oven to 375°F (190°C). Start by cooking 12 lasagna noodles according to package instructions. In a large skillet, brown 1 pound of ground beef with 1 diced onion and 2 minced garlic cloves. Drain any excess fat, then stir in 1 jar of marinara sauce and simmer for 10 minutes. In a separate bowl, mix 1 cup of ricotta cheese with 1 beaten egg and 1/2 cup of grated Parmesan cheese. To assemble the lasagna, spread a thin layer of meat sauce in the bottom of a baking dish. Layer 3 noodles over the sauce, then spread with a third of the ricotta mixture and sprinkle with shredded mozzarella cheese. Repeat the layers, ending with a final layer of meat sauce and mozzarella cheese. Cover with foil and bake for 25 minutes, then remove the foil and bake for an additional 15 minutes until the cheese is melted and bubbly. Let the lasagna cool for 10 minutes before slicing and serving.",
        },
        score: 1.5130847,
    },
    SearchResult {
        document: Document {
            id: "Macaroni and Cheese",
            contents: "To make classic macaroni and cheese, start by cooking 2 cups of elbow macaroni in a large pot of boiling salted water until al dente. Drain the pasta and set it aside. In the same pot, melt 2 tablespoons of butter over medium heat. Stir in 2 tablespoons of flour to form a roux, cooking for about 1 minute. Gradually whisk in 2 cups of milk, stirring constantly until the sauce thickens. Remove the pot from heat and stir in 2 cups of shredded cheddar cheese until melted and smooth. Add the cooked macaroni to the sauce, stirring to coat the pasta evenly. Serve hot, optionally topped with breadcrumbs and baked until golden.",
        },
        score: 1.0612293,
    },
    SearchResult {
        document: Document {
            id: "Pancakes",
            contents: "To make pancakes, whisk together 1 cup of flour, 2 tablespoons of sugar, 1 teaspoon of baking powder, and a pinch of salt in a large bowl. In a separate bowl, whisk together 1 cup of milk, 1 egg, and 2 tablespoons of melted butter. Pour the wet ingredients into the dry ingredients and stir until just combined; a few lumps are okay. Heat a non-stick skillet or griddle over medium heat and lightly grease it with butter or oil. Pour 1/4 cup of batter for each pancake onto the skillet. Cook until bubbles form on the surface, then flip and cook until the other side is golden brown. Serve warm with syrup and butter.",
        },
        score: 1.0612293,
    },
    SearchResult {
        document: Document {
            id: "Pumpkin Pie",
            contents: "To make pumpkin pie, preheat your oven to 425°F (220°C). In a large bowl, whisk together 2 cups of canned pumpkin puree, 1 can of sweetened condensed milk, 2 large eggs, 1 teaspoon of cinnamon, 1/2 teaspoon of ground ginger, 1/4 teaspoon of ground cloves, and 1/2 teaspoon of salt. Pour the mixture into an unbaked 9-inch pie crust. Bake for 15 minutes, then reduce the oven temperature to 350°F (175°C) and bake for an additional 35-40 minutes, or until a knife inserted into the center comes out clean. Let the pie cool completely before serving with whipped cream.",
        },
        score: 1.4620191,
    },
    SearchResult {
        document: Document {
            id: "Stuffed Bell Peppers",
            contents: "To make stuffed bell peppers, start by preheating your oven to 375°F (190°C). Cut the tops off 4 bell peppers and remove the seeds. In a large skillet, heat 1 tablespoon of olive oil over medium heat. Add 1 diced onion and 2 minced garlic cloves, cooking until softened. Stir in 1 pound of ground beef and cook until browned. Drain any excess fat. Add 1 cup of cooked rice, 1 can of diced tomatoes, salt, pepper, and a teaspoon of Italian seasoning. Stuff each bell pepper with the beef mixture and place them in a baking dish. Top with shredded cheese and cover with foil. Bake for 25-30 minutes, then remove the foil and bake for an additional 10 minutes until the cheese is bubbly and golden.",
        },
        score: 1.6142384,
    },
]
